core/app/models/spree/promotion/rules/user_role.rb

```diff
@@ -0,0 +1,26 @@
+module Spree
+  class Promotion
+    module Rules
+      class UserRole < PromotionRule
+        has_many :promotion_rule_roles,
+          class_name: 'Spree::PromotionRuleRole',
+          foreign_key: :promotion_rule_id,
+          dependent: :destroy
+
+        has_many :roles,
+          through: :promotion_rule_roles,
+          class_name: 'Spree::Role'
+
+        accepts_nested_attributes_for :roles
+
+        def applicable?(promotable)
+          promotable.is_a?(Spree::Order)
+        end
+
+        def eligible?(order, options = {})
+          (roles & order.user.spree_roles).any? if order.user
```

Comment:
You are right about the number of SQL queries; I didn't realize AR turned this into a JOIN query automatically.

With few roles you're right this not an issue. What happens if people start using this to make hundreds and hundreds of roles just because they want to segment specific promotions to specific people? Seems like 'danger zone' territory to me.